mltensor-explainer.md

```diff
@@ -0,0 +1,422 @@
+
+  // Return the buffer to WebNN.
+  tensorizedGpuBuffer.release();
+
+  // Perform some inference described by `graph` on the frame
+  // (e.g. selfie segmentation)
+  mlContext.dispatch(
+    graph,
+    /*inputs=*/{'input': mlTensor1},
+    /*outputs=*/{'output': mlTensor2},
+  );
+
+  // Wait for all ML work involving `mlTensor2` to complete, then rent it out to WebGPU.
+  const tensorizedGpuBufferAfterInference = await gpuDevice.importExternalBuffer(mlTensor2);
+
+  // Create a bind group for `tensorizedGpuBufferAfterInference`,
+  // create a command encoder, etc to feed `tensorizedGpuBufferAfterInference`
+  // into a GPU shader which may blur the frame or replace background sections
+  // and then render the result
+  // ...
+
+  gpuDevice.queue.submit([texturizeAndRenderCommandEncoder.finish()]);
+
+  // Return the buffer to WebNN for the next frame.
+  tensorizedGpuBufferAfterInference.release();
+
+  // Call this method for each frame.
+  video.requestVideoFrameCallback(applyEffectToFrame);
+}
+```
+
+## Design Discussion
+
+### Timelines
+
+WebNN uses a programming model similar to [WebGPU's](https://www.w3.org/TR/webgpu/#programming-model), in that compute tasks are posted to a timeline - which I've referred to as an "ML context timeline" throughout this document - separate from the content timeline (i.e. "script"). See [the WebGPU documentation of timelines](https://www.w3.org/TR/webgpu/#programming-model-timelines) for more details.
+
+Specifying WebNN timelines is tracked in [#529](https://github.com/webmachinelearning/webnn/issues/529).
+
+### Device Affinity and Relationship to a `GPUDevice`
+
+The WebNN API requires the developer to declare how an `MLTensor` will be used (via `MLTensorDescriptor`), which the user agent may use as a hint in deciding where to allocate the memory backing an `MLTensor`. Where the memory is ultimately allocated is up to the user agent.
+
+For example [an `MLContext` may be created with a `GPUDevice`](https://www.w3.org/TR/webnn/#dom-ml-createcontext-gpudevice), and creating an `MLTensor` from this context with the `MLTensorDescriptor.importableToWebGPU` flag expresses a clear intention to share the tensor with the given `GPUDevice`. However, there is no guarantee that sharing this tensor with WebGPU will be zero-copy.
+
+The `MLTensorDescriptor.readable` and `MLTensorDescriptor.writable` flags likewise are hints to the user agent indicating that the underlying data will be read and written to, respectively, by script.
+
+### Importing an `MLTensor` to WebGPU
+
+Any `MLTensor` created with the `MLTensorDescriptor.importableToWebGPU` flag may be imported into any `GPUDevice`. In the best case, this requires no data copies. If the underlying buffer backing the `MLTensor` is not accessible to the `GPUDevice`, this will require copying the contents of the `MLTensor` to a new buffer, then copying the contents of this buffer back to the `MLTensor` once WebGPU releases its handle to the buffer.
```

Comment:
How would this handle cases where the device buffer strides are not aligned with the tensor dimensions? Or the buffer storage data type is not the same as the tensor data type, i.e. is packed with quantized valued (4 int8 within a int32).

Reply:
> How would this handle cases where the device buffer strides are not aligned with the tensor dimensions? 

Hmm I would assume WGSL only cares about buffer strides, so the tensor dimensions would be irrelevant to WebGPU. Is there a scenario you have in mind?

> Or the buffer storage data type is not the same as the tensor data type,

Good question. Ideally the `importableToWebGPU` flag can be used as a hint to allocate the tensor in an import-friendly fashion, but at the end of the day I expect we'll have to make a data copy in many scenarios, and possibly also accept that some `MLTensor`s won't be importable to WebGPU at all. CoreML only supports creating an `MLMultiArray` with a handful of data types, for example, and [only float16 buffers can be imported to WebGPU without copies](https://developer.apple.com/documentation/coreml/mlmultiarray/init(pixelbuffer:shape:)). If it's not float16, a copy will be required on Mac

FWIW WebNN does not currently support packing, since no operators operate on packed data. That being said, folks at Intel are working on adding [a `dequantizeLinear` op which aims to support `"uint4"`](https://github.com/webmachinelearning/webnn/issues/375#issuecomment-2292466613) which will presumably require packing since there's no `Int4Array` in JS